VA-Explorer/va_explorer
# Issue 190: Callback conversions to clientside

Closes #190 

## Motivation
This PR attempts to speed up the loading of the dashboard by utilizing client-side callbacks instead of server-side callbacks when appropriate. The dashboard loading is not instantaneous and more data will only increase the amount of time it will take to load the data into the tool.

## Experiments & Results
To see whether client-side callbacks can actually speed up time, a few experiments were done. The first experiment examined the difference between the original server-side callbacks and the version of the code where the easiest callbacks were converted to client-side versions (for example, one of the callbacks just mapped the names of the tabs from something like "tab-1" to "COD Analysis"). There were four callbacks in total that were converted to client-side callbacks in this first experiment: 
- Dashboard Reset Callback (old function name: reset_dashboard) 
- Location Search Options (old function name: log_tab_value)
- Map metrics (old function name: get_metrics)
- Geographic Levels/View options (old function name: update_view_options)

The second experiment that was done examined the benefit of using Lodash instead of looping when we are working inside of the client-side callbacks (which are written in javascript). The use of Lodash did seem to slightly improve speed performance beyond using just loops. Compared to the old all-server-side version, there was a 8.33% improvement when using incorporating the client-side callbacks alone and an 8.56% improvement when using the client-side callbacks with Lodash. 

The final experiment looked at converting the callout boxes callback, which turned out to be a bit more complicated to convert due to having a few more connections to items on the server-side (namely the use of some functions to create the html objects that would display the callout boxes as well as references to global variables). After taking the time to successfully convert this callback to do more computations on the client-side, the speed of loading the dashboard decreased compared the previous version that utilized Lodash. The experiment found, across 20 page refreshes, that the all-server-side code took an average of 7.34 seconds to load the page, the Lodash with the preliminary client-side conversions took an average of 6.72 seconds to load the page, while the version that incorporated this final client-side converted callback took an average of 7.03 seconds. (Note that these times were recorded when the development toolbar was activated, which slows things down a little bit). The results indicate that converting the callout boxes callback to be a client-side callback increased the average time to load the page compared to leaving it as a server-side only callback. 


## Other miscellaneous changes made when addressing this issue
- Bug fix relating to the Trend Plot of the VA Trends tab of the Dashboard: it should now update its data and title when you select a new item from the Cause of Death dropdown towards the top of the page (Note: this is _not_ referring to the "Search COD Keywords" search bar but rather dropdown next to the Time Period dropdown and Reset button)
- Uncommented a line in plotting.py that was causing an issue with loading times when changing back and forth between different items in the Time Period dropdown. Previously, you could quickly change the Time Period and the dashboard would quickly load the new data, but changing it a second time would cause long waiting times.

## Testing Instructions
- The dashboard functionality should be exactly as it was before
- There should be an improvement in the speed of loading the new version of va_dashboard.py compared to the previous version.
    - In Microsoft Edge, you can check this by right-clicking your webpage, clicking "Inspect Page" and then finding "Network" in the tabs at the very top of the popout (you may have to check under  >> to see additional options). If you refresh the page you should be able to find how long it took about halfway down the page in the format of: "Finish : X.XX s".
- The data and title of the plot in the "VA Trends" Tab should update to when selecting a new item in the Cause of Death dropdown
    - Change the selection of the Cause of Death dropdown and note that the data and title changed. The title should no longer say "**All Causes** by Month" but "<selected COD here> by Month"
- There should no longer be a lag when switching between items in the Time Period dropdown:
    - Try changing the Time Period a few times and making sure that the data in the map on the left and plots on the right are updating in a reasonable time span (a few seconds). You shouldn't see 30 second+ wait times when changing the dropdown the second time.
